tests/test_protocol_object_conversion.py

```diff
@@ -1,67 +1,81 @@
+from typing import Type, TypeVar, Union
+
 import pytest
 
 from aiokafka.protocol.admin import Request, Response
 from aiokafka.protocol.types import Array, Int16, Schema, String
 
+C = TypeVar("C", bound=Type[Union[Request, Response]])
 
-@pytest.mark.parametrize("superclass", (Request, Response))
-class TestObjectConversion:
-    def test_get_item(self, superclass):
-        class TestClass(superclass):
+
+def _make_test_class(klass: C, schema: Schema) -> C:
+    if issubclass(klass, Request):
+
+        class TestClass(Request):
+            API_KEY = 0
+            API_VERSION = 0
+            RESPONSE_TYPE = Response
+            SCHEMA = schema
+
+    else:
+
+        class TestClass(Response):  # type: ignore[no-redef]
             API_KEY = 0
             API_VERSION = 0
-            RESPONSE_TYPE = None  # To satisfy the Request ABC
-            SCHEMA = Schema(("myobject", Int16))
+            SCHEMA = schema
+
+    return TestClass  # type: ignore[return-value]
```

Comment:
Hmm, looks like mypy is right here, as the type we return is not the same. I don't know how to solve this in generic way, but we actually don't need to. Also, it's incorrect to use `issubclass` here, as we accept only `Request` and `Response`, but not subclasses.  Maybe this way?:

```python
def _make_test_class(base_cls: Type[Union[Request, Response]]) ->  Type[Union[Request, Response]]:
    if base_cls is Request:
        class TestRequest(Request):
            ...

        return TestRequest

    else:
        assert base_cls is Response

        class TestResponse(Response):
            ...

        return TestResponse
```

Reply:
I prefer generic because if we add annotations to the __init__ method, we should know the return value from class factory

Reply:
Maybe use the version with generic (and # type: ignore[return-value]) and add a protocol like this? This solves the problem of differences between Response/Request.

```python
class ResponseRequestProtocol(Protocol):
    def __init__(self, *args: Any, **kwargs: Any) -> None: ...
    def to_object(self) -> Dict[str, Any]: ...
    def encode(self) -> bytes: ...
    def get_item(self, name: str) -> Any: ...
```

Reply:
> I prefer generic because if we add annotations to the init method, we should know the return value from class factory

I don't know how to express this with generic. It's not very good, but we can use `type: ignore` if we're sure our typing is right and there is a bug in mypy. But here I'm pretty sure that mypy is right and typing is incorrect: when `TestClass` is subclass of `Request` (or `Response`) this doesn't mean that `Type[TestClass]` is subclass of `Type[Request]` (or `Type[Response]`). And don't see how suggested protocol may solve this problem.

Reply:
BTW, do you have a Telegram account? Mine is [otkds](https://t.me/otkds), probably we might discuss ideas in realtime. (I'll remove this comment later to avoid spam)